### PR TITLE
LMS/Studio give 404 if user has no access to request site

### PR DESCRIPTION
**Pull request by asadmanzoor93**
**Description:**  This PR adds the following implementation:

- Add middleware to raise Http404 if the request site is not linked to the user’s edly sub-organization.

- Add Unit Tests for Middleware.

**JIRA:** 
https://edlyio.atlassian.net/browse/EDLY-1322

**Merge checklist:**

- [ ] All reviewers approved
- [ ] Commits are (reasonably) squashed

**Post merge:**
- [ ] Delete working branch (if not needed anymore)

**Note:** Please add screenshots for any viewable change.

### CODE REVIEW

File: openedx/features/edly/utils.py [hunk 1]
```diff
@@ -25,20 +25,24 @@ def user_has_edly_organization_access(request):
...
-    ).first()
+    try:
+        if not request.user.edly_profile:
+            return False
+
+        current_site = request.site
+        edly_sub_org = EdlySubOrganization.objects.filter(
+            Q(lms_site=current_site) |
+            Q(studio_site=current_site)
+        ).first()
```
Comment (rehan99000) [inline openedx/features/edly/utils.py:37]: This can be done as
```
edly_sub_org = EdlySubOrganization.objects.get(
    Q(lms_site=current_site) |
    Q(studio_site=current_site)
)
```
with exception `DoesNotExist`

Comment (asadmanzoor93) [reply openedx/features/edly/utils.py]: Feedback Implemented

File: openedx/features/edly/middleware.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    """
+    Django middleware to validate edly user organization access based on request.
+    """
+
+    def process_request(self, request):
+        """
+        Validate edly user organization access based on request.
+        """
+
+        if not user_has_edly_organization_access(request) and request.user.is_authenticated:
```
Comment (zubair-arbi) [inline openedx/features/edly/middleware.py:22]: first check if use is authenticated then check the user access.
```python
if request.user.is_authenticated and not user_has_edly_organization_access(request):
```

File: openedx/features/edly/tests/test_middleware.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        self.request = RequestFactory().get('/')
+        self.request.user = self.user
+        self.request.site = SiteFactory()
+
+        self.client = Client(SERVER_NAME=self.request.site.domain)
+        self.client.login(username=self.user.username, password='test')
+
+    def _create_edly_sub_organization(self):
+        """
+        Helper method to create 'EdlySubOrganization` for the request site.
```
Comment (zubair-arbi) [inline openedx/features/edly/tests/test_middleware.py:25]: nit: `"EdlySubOrganization"`

File: openedx/features/edly/tests/test_middleware.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        """
+        return EdlySubOrganizationFactory(lms_site=self.request.site)
+
+    def test_user_with_organization_access(self):
+        """
+        Test edly user organization with access based on request.
```
Comment (zubair-arbi) [inline openedx/features/edly/tests/test_middleware.py:31]: Test logged in user access based on user's linked edly sub organization.

File: openedx/features/edly/tests/test_middleware.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        self.client = Client(SERVER_NAME=self.request.site.domain)
+        self.client.login(username=self.user.username, password='test')
+
+    def _create_edly_sub_organization(self):
+        """
+        Helper method to create 'EdlySubOrganization` for the request site.
+        """
+        return EdlySubOrganizationFactory(lms_site=self.request.site)
+
+    def test_user_with_organization_access(self):
```
Comment (zubair-arbi) [inline openedx/features/edly/tests/test_middleware.py:29]: `test_user_with_edly_organization_access`

File: openedx/features/edly/tests/test_middleware.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        self.client.cookies.load(
+            {
+                settings.EDLY_USER_INFO_COOKIE_NAME: cookies._get_edly_user_info_cookie_string(self.request)
+            }
+        )
+
+        response = self.client.get('/', follow=True)
+        assert response.status_code == 200
+
+    def test_user_without_organization_access(self):
```
Comment (zubair-arbi) [inline openedx/features/edly/tests/test_middleware.py:43]: `test_user_without_edly_organization_access`

File: openedx/features/edly/tests/test_middleware.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
+"""
+Unit tests for Middleware.
+"""
+from django.conf import settings
+from django.test import TestCase
+from django.test.client import Client, RequestFactory
+
+from openedx.features.edly import cookies
+from openedx.features.edly.tests.factories import EdlySubOrganizationFactory, EdlyUserFactory, SiteFactory
```
Comment (iam-mhaseeb) [inline openedx/features/edly/tests/test_middleware.py:9]: Python's import style prefer that!
```
from openedx.features.edly.tests.factories import (
EdlySubOrganizationFactory, 
EdlyUserFactory, 
SiteFactory
)
```

File: openedx/features/edly/tests/test_middleware.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+                settings.EDLY_USER_INFO_COOKIE_NAME: cookies._get_edly_user_info_cookie_string(self.request)
+            }
+        )
+
+        response = self.client.get('/', follow=True)
+        assert response.status_code == 200
+
+    def test_user_without_organization_access(self):
+        """
+        Test edly user organization without access based on request.
```
Comment (zubair-arbi) [inline openedx/features/edly/tests/test_middleware.py:45]: Test that logged in user gets 404 if user has no access for request site's edly sub organization.

File: openedx/features/edly/middleware.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class EdlyOrganizationAccessMiddleware(object):
+    """
+    Django middleware to validate edly user organization access based on request.
+    """
+
+    def process_request(self, request):
+        """
+        Validate edly user organization access based on request.
```
Comment (zubair-arbi) [inline openedx/features/edly/middleware.py:19]: `Validate logged in user's access based on request site and its linked edly sub organization.`

File: openedx/features/edly/middleware.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+        """
+
+        if not user_has_edly_organization_access(request) and request.user.is_authenticated:
+            logger.warning('Edly user organization does not have access')
```
Comment (zubair-arbi) [inline openedx/features/edly/middleware.py:23]: `Edly user %s has no access for site %s.` % (user email, request.site)

File: openedx/features/edly/utils.py [hunk 11]
```diff
@@ -25,20 +25,24 @@ def user_has_edly_organization_access(request):
...
     """
-    if not request.user.edly_profile:
-        return False
 
-    current_site = request.site
-    edly_sub_org = EdlySubOrganization.objects.filter(
-        Q(lms_site=current_site) |
-        Q(studio_site=current_site)
-    ).first()
+    try:
```
Comment (zubair-arbi) [inline openedx/features/edly/utils.py:29]: Revert these changes back

File: openedx/features/edly/tests/test_middleware.py (hunk 7 above)
Comment (asadmanzoor93) [reply openedx/features/edly/tests/test_middleware.py:9]: Isort is removing that since it's not violating length of the line, however, I will add this explicitly

File: openedx/features/edly/utils.py (hunk 11 above)
Comment (rehan99000) [reply openedx/features/edly/utils.py:29]: @zubair-arbi since there is a `OneToOne` relation of site and `EdlySubOrganization`, shouldn't we `get` it instead of filtering?

File: openedx/features/edly/tests/test_middleware.py (hunk 7 above)
Comment (zubair-arbi) [reply openedx/features/edly/tests/test_middleware.py]: Don't forget to put comma with the last element `SiteFactory,`

File: openedx/features/edly/utils.py (hunk 11 above)
Comment (zubair-arbi) [reply openedx/features/edly/utils.py:29]: Yes, that's fine

File: openedx/features/edly/utils.py [hunk 12]
```diff
@@ -25,14 +25,19 @@ def user_has_edly_organization_access(request):
...
+        edly_sub_org = EdlySubOrganization.objects.get(
+            Q(lms_site=current_site) |
+            Q(studio_site=current_site)
+        )
+    except EdlySubOrganization.DoesNotExist:
+        return False
 
     edly_user_info_cookie = request.COOKIES.get(settings.EDLY_USER_INFO_COOKIE_NAME, None)
     if edly_sub_org and edly_sub_org.slug == get_edly_sub_org_from_cookie(edly_user_info_cookie):
```
Comment (zubair-arbi) [inline openedx/features/edly/utils.py:43]:
```python
if edly_sub_org.slug == get_edly_sub_org_from_cookie(edly_user_info_cookie):
```

File: openedx/features/edly/middleware.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    Django middleware to validate edly user organization access based on request.
+    """
+
+    def process_request(self, request):
+        """
+        Validate logged in user's access based on request site and its linked edly sub organization.
+        """
+
+        if request.user.is_authenticated and not user_has_edly_organization_access(request):
+            logger.warning('Edly user %s has no access for site %s.' % (request.user.email, request.site))
```
Comment (zubair-arbi) [inline openedx/features/edly/middleware.py:23]: Please add test for the log message.

File: openedx/features/edly/middleware.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
+"""
+Edly Organization Access Middleware
```
Comment (zubair-arbi) [inline openedx/features/edly/middleware.py:2]: nit: `Edly Organization Access Middleware.`

Comment (asadmanzoor93) [reply openedx/features/edly/middleware.py]: Done

File: openedx/features/edly/middleware.py (hunk 13 above)
Comment (asadmanzoor93) [reply openedx/features/edly/middleware.py]: Unit Test updated to check log message on 404.

File: openedx/features/edly/utils.py (hunk 12 above)
Comment (asadmanzoor93) [reply openedx/features/edly/utils.py]: Done

File: openedx/features/edly/middleware.py [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+    """
+    Django middleware to validate edly user organization access based on request.
+    """
+
+    def process_request(self, request):
+        """
+        Validate logged in user's access based on request site and its linked edly sub organization.
+        """
+
+        if request.user.is_authenticated and not user_has_edly_organization_access(request):
```
Comment (zubair-arbi) [inline openedx/features/edly/middleware.py:22]: First we need to check if the `request.user` is superuser then we need to allow the user to proceed with the request. Something like (please verify for anonymous users):

```python
user_is_authenticated = request.user.is_authenticated
user_is_superuser = request.user.is_superuser)
if user_is_authenticated and not user_is_superuser and not user_has_edly_organization_access(request)):
```

So that the MAU's report request which is requested on behalf of superuser don't get affect.

Please add related test as well.

Comment (rehan99000) [reply openedx/features/edly/middleware.py:22]: Good catch!

Comment (asadmanzoor93) [reply openedx/features/edly/middleware.py]: Condition for `Super User` in Middleware and relevant Unit Test added.

File: openedx/features/edly/tests/test_middleware.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+            logger.check(
+                (
+                    LOGGER_NAME,
+                    'ERROR',
+                    'Edly user %s has no access for site %s.' % (self.user.email, self.request.site)
+                )
+            )
+
+            assert response.status_code == 404
```
Comment (zubair-arbi) [inline openedx/features/edly/tests/test_middleware.py:69]: nit: first check the response code then check logger
```python
response = self.client.get('/', follow=True)
assert response.status_code == 404
logger.check(...)
```

File: openedx/features/edly/tests/test_middleware.py [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    def setUp(self):
+        self.user = EdlyUserFactory()
+        self.request = RequestFactory().get('/')
+        self.request.user = self.user
+        self.request.site = SiteFactory()
+
+        self.client = Client(SERVER_NAME=self.request.site.domain)
+        self.client.login(username=self.user.username, password='test')
+
+    def _create_edly_sub_organization(self):
```
Comment (zubair-arbi) [inline openedx/features/edly/tests/test_middleware.py:30]: remove this method as its just one liner and also its used only once in this test class.

File: openedx/features/edly/tests/test_middleware.py (hunk 16 above)
Comment (asadmanzoor93) [reply openedx/features/edly/tests/test_middleware.py]: Done

File: openedx/features/edly/tests/test_middleware.py (hunk 17 above)
Comment (asadmanzoor93) [reply openedx/features/edly/tests/test_middleware.py]: Done
